% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{booktabs}
\usepackage[utf8]{inputenc}
\usepackage{geometry}
\geometry{a4paper,left=2.5cm,right=2.5cm,top=2.5cm,bottom=2.5cm}
\usepackage{amsmath,amssymb,amsfonts}
\usepackage[colorinlistoftodos]{todonotes}
\usepackage{algorithmic}
\usepackage{graphicx}
\usepackage{subfigure}
\usepackage{textcomp}
\usepackage{xcolor}
\usepackage[version=4]{mhchem}
\usepackage{multirow}
\usepackage{booktabs}
\usepackage{threeparttable}
\usepackage{float}
\usepackage{textcomp}
\usepackage{bm}
\usepackage{authblk}
\usepackage{listings}
\usepackage{fvextra}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaklines,commandchars=\\\{\}}
\lstset{
  basicstyle=\ttfamily,
  columns=fullflexible,
  frame=single,
  breaklines=true,
  postbreak=\mbox{\textcolor{red}{$\hookrightarrow$}\space},
}

%



\ifluatex
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage[]{natbib}
\bibliographystyle{plainnat}

\author{}
\date{\vspace{-2.5em}}

\begin{document}

\author[1]{Kejie Li}
\author[2]{Zhengyu Ouyang}
\author[1]{Yirui Chen}
\author[1]{Dongdong Lin}
\author[1]{Michael Mingueneau}
\author[1]{Will Chen}
\author[1]{David Sexton}
\author[1]{Baohong Zhang\thanks{corresponding author: baohong.zhang@biogen.com}}
\affil[1]{Research Department, Biogen, Inc., 225 Binney St, Cambridge, MA 02142, USA }
\affil[2]{BioInfoRx, Inc., 510 Charmany Dr, Suite 275A, Madison, WI 53719, USA}
                                           
\title{\textbf{Supplementary Materials}: cellxgene VIP unleashes full power of interactive visualization, plotting and analysis of scRNA-seq data in the scale of millions of cells}

\maketitle


{
\setcounter{tocdepth}{2}
\tableofcontents
}
\hypertarget{getting-started-with-cellxgene-vip}{%
\section{Getting started with cellxgene VIP}\label{getting-started-with-cellxgene-vip}}

This is a cellxgene VIP tutorial book written in \textbf{Markdown}.

\hypertarget{why-use-cellxgene-vip}{%
\subsection{Why use cellxgene VIP?}\label{why-use-cellxgene-vip}}

To meet the growing demands from scientists to effectively extract deep insights from single cell RNA-seq datasets, we developed cellxgene VIP, a frontend interactive visualization plugin to cellxgene framework, which directly interacts with in-memory data to generate a comprehensive set of plots in high resolution, perform advanced analysis, and make data downloadable for further analysis. It makes large scale scRNA-seq data visualization and analysis more accessible and reproducible with the potential to become an ecosystem for the scientific community to contribute even more modules to the Swiss knife of scRNA-seq data exploration tool.

\hypertarget{getting-set-up}{%
\subsection{Getting Set up}\label{getting-set-up}}

\hypertarget{execute-anaconda}{%
\subsubsection{Execute anaconda}\label{execute-anaconda}}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{bash}\NormalTok{ \textasciitilde{}/Downloads/Anaconda3{-}2020.02{-}Linux{-}x86\_64.sh}
\end{Highlighting}
\end{Shaded}

If anaconda is not installed on server, you can install it following anaconda documentation (\url{https://docs.anaconda.com/anaconda/install/linux/})
\#\#\# Create and enable conda environment

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# clone repo from cellxgene VIP github}
\FunctionTok{git}\NormalTok{ clone https://github.com/interactivereport/cellxgene\_VIP.git}
\BuiltInTok{cd}\NormalTok{ cellxgene\_VIP}

\CommentTok{\# conda environment}
\BuiltInTok{source} \OperatorTok{\textless{}}\NormalTok{path to Anaconda3}\OperatorTok{\textgreater{}}\NormalTok{/etc/profile.d/conda.sh }\ErrorTok{(}\ExtensionTok{Default:}\NormalTok{ /opt/anaconda3/etc/profile.d/conda.sh}\KeywordTok{)}
\ExtensionTok{conda}\NormalTok{ config }\AttributeTok{{-}{-}set}\NormalTok{ channel\_priority flexible}
\ExtensionTok{conda}\NormalTok{ env create }\AttributeTok{{-}n} \OperatorTok{\textless{}}\NormalTok{env name, such as: VIP}\OperatorTok{\textgreater{}}\NormalTok{ {-}f VIP.yml }\ErrorTok{(}\ExtensionTok{system{-}wide}\NormalTok{ R}\KeywordTok{)} \ExtensionTok{or}\NormalTok{ VIP\_conda\_R.yml }\ErrorTok{(}\BuiltInTok{local} \VariableTok{R} \VariableTok{under} \VariableTok{conda}\NormalTok{, }\VariableTok{no} \VariableTok{root} \VariableTok{privilege} \VariableTok{needed}\KeywordTok{)}
\end{Highlighting}
\end{Shaded}

Activate conda environment

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{conda}\NormalTok{ activate }\OperatorTok{\textless{}}\NormalTok{env name, such as: VIP}\OperatorTok{\textgreater{}}
\end{Highlighting}
\end{Shaded}

or

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{source}\NormalTok{ activate }\OperatorTok{\textless{}}\NormalTok{env name}\OperatorTok{\textgreater{}}
\end{Highlighting}
\end{Shaded}

\hypertarget{cellxgene-installation}{%
\subsubsection{Cellxgene installation}\label{cellxgene-installation}}

Install cellxgene by running config.sh in ``cellxgene\_VIP'' directory

\begin{Shaded}
\begin{Highlighting}[]
\ExtensionTok{./config.sh}
\end{Highlighting}
\end{Shaded}

\hypertarget{r-dependencies}{%
\subsubsection{R dependencies}\label{r-dependencies}}

Install all required R packages on linux:

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{export} \VariableTok{LIBARROW\_MINIMAL=}\NormalTok{false}
\CommentTok{\#  ensure that the right instance of R is used. e.g. system{-}wide: /bin/R or /usr/bin/R ; local R under conda: \textasciitilde{}/.conda/envs/VIP\_conda\_R/bin/R}
\FunctionTok{which}\NormalTok{ R}

\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(devtools)) install.packages("devtools",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(Cairo)) devtools::install\_version("Cairo",version="1.5{-}12",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(foreign)) devtools::install\_version("foreign",version="0.8{-}76",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(ggpubr)) devtools::install\_version("ggpubr",version="0.3.0",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(ggrastr)) devtools::install\_version("ggrastr",version="0.1.9",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(arrow)) devtools::install\_version("arrow",version="2.0.0",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(Seurat)) devtools::install\_version("Seurat",version="3.2.3",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(rmarkdown)) devtools::install\_version("rmarkdown",version="2.5",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(tidyverse)) devtools::install\_version("tidyverse",version="1.3.0",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(viridis)) devtools::install\_version("viridis",version="0.5.1",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(BiocManager)) devtools::install\_version("BiocManager",version="1.30.10",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(fgsea)) BiocManager::install("fgsea")\textquotesingle{}}

\CommentTok{\# These should be already installed as dependencies of above packages}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(dbplyr)) devtools::install\_version("dbplyr",version="1.0.2",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(RColorBrewer)) devtools::install\_version("RColorBrewer",version="1.1{-}2",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(glue)) devtools::install\_version("glue",version="1.4.2",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(gridExtra)) devtools::install\_version("gridExtra",version="2.3",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(ggrepel)) devtools::install\_version("ggrepel",version="0.8.2",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(MASS)) devtools::install\_version("MASS",version="7.3{-}51.6",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\ExtensionTok{R} \AttributeTok{{-}q} \AttributeTok{{-}e} \StringTok{\textquotesingle{}if(!require(data.table)) devtools::install\_version("data.table",version="1.13.0",repos = "http://cran.us.r{-}project.org")\textquotesingle{}}
\end{Highlighting}
\end{Shaded}

\hypertarget{run-cellxgene-by-h5ad-file}{%
\subsubsection{Run cellxgene by h5ad file}\label{run-cellxgene-by-h5ad-file}}

You can aslo run cellxgene by specifying a h5ad file, which stores scRNA-seq data along with a host and a port.
Use `ps' to find used ports to spare. Please see \url{https://chanzuckerberg.github.io/cellxgene/posts/launch} for details

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{ps} \AttributeTok{{-}ef} \KeywordTok{|} \FunctionTok{grep}\NormalTok{ cellxgene}
\ExtensionTok{Rscript} \AttributeTok{{-}e} \StringTok{\textquotesingle{}reticulate::py\_config()\textquotesingle{}}
\CommentTok{\# Run the following command if the output of the above command doesn\textquotesingle{}t point to the Python in your env.}
\BuiltInTok{export} \VariableTok{RETICULATE\_PYTHON=}\KeywordTok{\textasciigrave{}}\FunctionTok{which}\NormalTok{ python}\KeywordTok{\textasciigrave{}}
\ExtensionTok{cellxgene}\NormalTok{ launch }\AttributeTok{{-}{-}host} \OperatorTok{\textless{}}\NormalTok{xxx}\OperatorTok{\textgreater{}}\NormalTok{ {-}{-}port }\OperatorTok{\textless{}}\NormalTok{xxx}\OperatorTok{\textgreater{}}\NormalTok{ {-}{-}disable{-}annotations }\AttributeTok{{-}{-}verbose} \OperatorTok{\textless{}}\NormalTok{h5ad file}\OperatorTok{\textgreater{}}
\end{Highlighting}
\end{Shaded}

\hypertarget{cellxgene-on-web-browser}{%
\subsubsection{Cellxgene on web browser}\label{cellxgene-on-web-browser}}

chrome is preferred, version 87.0.4280.88 or 87.0.4280.141 is used. Users can access \textbf{http(s)://host:port}.
Following screenshot is what you should be able to see in console of chrome developer tools.
\includegraphics{cellonweb.png}

\hypertarget{authors}{%
\subsection{Authors}\label{authors}}

\begin{itemize}
\tightlist
\item
  Keijie Li (\href{mailto:kejie.li@biogen.com}{\nolinkurl{kejie.li@biogen.com}}), Associate Director at Biogen in the research department. Main author and content wrangler.
\item
  Zhengyu Ouyang, Associate Director of Bioingormatics at BioinfoRx. Content wrangler.
\item
  Baohong Zhang (\href{mailto:baohong.zhang@biogen.com}{\nolinkurl{baohong.zhang@biogen.com}}), Head of Genome Informatics at Biogen in the research department. Corresponding author and content wrangler.
\end{itemize}

The development and delivery of this material has also contributed by:

\begin{itemize}
\tightlist
\item
  Yirui Chen (\href{mailto:yirui.chen@bigen.com}{\nolinkurl{yirui.chen@bigen.com}}), Biogen Inc.
\item
  Dongdong Lin (\href{mailto:dongdong.lin@biogen.com}{\nolinkurl{dongdong.lin@biogen.com}}), Biogen Inc.
\item
  Michael Mingueneau (\href{mailto:michael.mingueneau@bioigen.com}{\nolinkurl{michael.mingueneau@bioigen.com}}), Biogen Inc.
\item
  Will Chen (\href{mailto:wwchen@post.harvard.edu}{\nolinkurl{wwchen@post.harvard.edu}}), Biogen Inc.
\item
  David Sexton (\href{mailto:david.sexton@biogen.com}{\nolinkurl{david.sexton@biogen.com}}), Biogen Inc.
\end{itemize}

\hypertarget{web-source}{%
\section{Web source}\label{web-source}}

\hypertarget{cellxgene-vip}{%
\subsection*{Cellxgene VIP}\label{cellxgene-vip}}
\addcontentsline{toc}{subsection}{Cellxgene VIP}

\begin{itemize}
\tightlist
\item
  \textbf{cellxgene VIP source code}: \url{https://github.com/interactivereport/cellxgene_VIP}
\item
  \textbf{cellxgene VIP demo sites}: \url{https://cellxgenevip-ms.bxgenomics.com} {[}Schirmer / Rowitch MS, Human brain snRNAseq 46k cells, Nature 2019 Schirmer et al{]}
\end{itemize}

\hypertarget{cellxgene}{%
\subsection*{Cellxgene}\label{cellxgene}}
\addcontentsline{toc}{subsection}{Cellxgene}

\begin{itemize}
\tightlist
\item
  \textbf{cellxgene}: \url{https://github.com/chanzuckerberg/cellxgene}
\item
  \textbf{cellxgene tutorial}: \url{https://cellgeni.readthedocs.io/en/latest/visualisations.html}
\item
  \textbf{cellxgene features}: \url{https://chanzuckerberg.github.io/cellxgene/posts/gallery}
\item
  \textbf{cellxgene data preparation}: \url{https://chanzuckerberg.github.io/cellxgene/posts/prepare.html}
\end{itemize}

\hypertarget{python-packages}{%
\subsection*{Python packages}\label{python-packages}}
\addcontentsline{toc}{subsection}{Python packages}

\begin{itemize}
\tightlist
\item
  \textbf{scanpy}: \url{https://github.com/theislab/scanpy}
\item
  \textbf{scanpy plots}: \url{https://scanpy-tutorials.readthedocs.io/en/latest/visualizing-marker-genes.html}
\item
  \textbf{diffxpy}: \url{https://github.com/theislab/diffxpy}
\item
  \textbf{diffxpy tests}: \url{https://diffxpy.readthedocs.io/en/latest/tutorials.html\#differential-testing}
\end{itemize}

\hypertarget{others}{%
\subsection*{Others}\label{others}}
\addcontentsline{toc}{subsection}{Others}

\begin{itemize}
\tightlist
\item
  \textbf{Benchmarking of interactive data visualization of single-cell RNAseq data --- Batuhan Cakir} : \url{https://www.youtube.com/watch?v=3nH2xi_Ni6I}
\item
  \textbf{sceasy}: convertor of scRNA-Seq data formats \url{https://github.com/cellgeni/sceasy}
\item
  \textbf{jupytext}: \url{https://jupytext.readthedocs.io} , Jupyter Notebooks as Markdown Documents, Julia, Python or R Scripts
\item
  \textbf{nbconvert}: \url{https://nbconvert.readthedocs.io} , Convert a Jupyter .ipynb notebook document file into another static format including HTML, LaTeX, PDF, Markdown, and more
\end{itemize}

\hypertarget{how-to-use-cellxgene-vip}{%
\section{How to use Cellxgene VIP}\label{how-to-use-cellxgene-vip}}

\hypertarget{graphical-user-interface-of-cellxgene-and-vip}{%
\subsection{Graphical user interface of cellxgene and VIP}\label{graphical-user-interface-of-cellxgene-and-vip}}

The main window of cellxgene is divided into three regions, the left panel mainly displays categorial
annotations, brief description of the data set and initial graphics setting, specifically embedding and
coloring of cells. On the right panel, it hosts continuous variables, such as qc metrics shown in histogram
with x, y corresponding to values of a measurement and numbers of cells, respectively. More
importantly, cells shown as individual dots are presented in the center panel based on a selected
embedding and colored by either categorial annotations or continuous variables, which is indicated by
pressed rain drop icon.

\begin{figure}
\centering
\includegraphics{figures/F1A.jpg}
\caption{Cellxgene main window, functional icons and minimized VIP bar next to cellxgene logo}
\end{figure}

\begin{figure}
\centering
\includegraphics{figures/F1B.jpg}
\caption{VIP (Visualization in Plugin) window and controls of user interface. The cursor will change to corresponding icon when mouse hovers over control anchors inside the window. In the case of missing title bar after operation, changing the size of outside browser window (not VIP window) will always bring the VIP window back to the original location near the cellxgene logo}
\end{figure}

\hypertarget{cell-selection-by-categorial-annotations}{%
\subsection{Cell selection by categorial annotations}\label{cell-selection-by-categorial-annotations}}

\begin{figure}
\centering
\includegraphics{figures/F2.jpg}
\caption{Cell selection by categorial annotation. Selected B cells are shown in bold dots and highlighted in purple color when hovering mouse over the cluster}
\end{figure}

It is an overlap operation when categories from multiple annotations are checked to make the final selection. E.g., if male from sex is also checked besides B cells, it means cells from B cells cluster of male samples are selected.
Note: Click ``1:'' or ``2:'' button to save cell selection into group 1 or 2

\hypertarget{cell-selection-by-brushing-on-distribution-of-continues-variables}{%
\subsection{Cell selection by brushing on distribution of continues variables}\label{cell-selection-by-brushing-on-distribution-of-continues-variables}}

\begin{figure}
\centering
\includegraphics{figures/F3.jpg}
\caption{Cell selection by brushing the ranges of continuous variables. Low- and high-end values are shown at top corners of brushing boxes in dark gray}
\end{figure}

Note: Histograms of expression values of genes can by brushed as well to get cells expressing certain genes in the range.

\hypertarget{free-hand-lasso-selection-on-dots-representing-cells}{%
\subsection{Free hand Lasso selection on dots representing cells}\label{free-hand-lasso-selection-on-dots-representing-cells}}

From the cell visualization panel, user can freely select a cluster of cells of interest by using `Lasso' selection tool. The selected cluster of cells can also be added as a group for downstream analysis in cellxgene VIP.

\begin{figure}
\centering
\includegraphics{figures/F4.jpg}
\caption{Select cells by using free hand Lasso selection tool and add these cells as a group for further analysis in cellxgene VIP}
\end{figure}

Note: Please try to draw as close as possible to the starting point in the end to make an enclosed shape to ensure successfully lasso selection.

\hypertarget{vip-figure-option}{%
\subsection{VIP -- Figure Option}\label{vip-figure-option}}

User can set parameters for figure plotting that control plotting functions except CLI. `split\_show' branch of Scanpy offers better representation of Stacked Violin and Dot Plot comparing to master branch.

\begin{figure}
\centering
\includegraphics{figures/F5.jpg}
\caption{Setting parameters for figure plotting}
\end{figure}

Scaled data have zero mean and unit variance per gene. This was performed by calculating z-scores of the expression data using Scanpy's scale function. (Scanpy pp.scale function: Scale data to unit variance and zero mean.)

We provide flexibility to allow 1) scale to unit variance or not; 2) Zero centered or not; 3) Capped at max value after scaling.

We recommend using scaled data for plotting/visualization while using non-scaled data for differential gene expression analysis.

Note: Dot plot is one exception in visualization category which uses non-scaled data for meaningful interpretation.

\hypertarget{vip-add-genes-gene-sets}{%
\subsection{VIP -- Add Genes / Gene Sets}\label{vip-add-genes-gene-sets}}

Cellxgene VIP allows user to add any genes or gene sets for extensive exploration and visualization. User can either type a list of gene in the textbox or create sets of genes to be grouped together in plots. Then the genes will be automatically listed for plotting in other functional modules after checking availability in the dataset.

\begin{figure}
\centering
\includegraphics{figures/F6.jpg}
\caption{Add gene or gene sets for plotting}
\end{figure}

Note: The cursor will turn to cross icon while hovering over a gene name, then click to delete the gene.

\hypertarget{vip-violin-plot}{%
\subsection{VIP -- Violin Plot}\label{vip-violin-plot}}

To plot expression of gene among categories of an annotation, e.g., cell type, sex, or batch etc.
Step 1. User needs to select the group(s) of cells for plotting. These groups can be created by using selection tools illustrated in tutorial section 2, 3 and/or 4. Initially, all of cells are gathered in `Group 1' by default.
Step 2. Select a gene from the gene list which could be added as shown in section 6. An expression level cutoff can be set to further filter out cells with low level expression of such gene.
Step 3. Select the annotation to group cells for plotting.
Step 4. Execute plotting, get plotting data (i.e., gene expression), manipulate image (e.g., zoom in/out) or save the image.

\includegraphics{figures/F7.jpg}
Note: Figure resolution and format can be set in ``Figure Option'' tab as shown in tutorial section 5.

\hypertarget{vip-stacked-violin}{%
\subsection{VIP -- Stacked Violin}\label{vip-stacked-violin}}

Beyond plotting expression values of a gene, stacked violin allows plotting of multiple genes together.

\begin{figure}
\centering
\includegraphics{figures/F8.jpg}
\caption{Stacked Violin plot of multiple genes and/or gene set}
\end{figure}

Note: If collapsing of gene sets is set to `Yes', average gene expression of genes in a set is used for plotting.

\hypertarget{vip-heatmap}{%
\subsection{VIP -- Heatmap}\label{vip-heatmap}}

To show or compare the expression level (i.e., expression value or expression Z-score) of multiple genes among the selected group of cells.

\begin{figure}
\centering
\includegraphics{figures/F9.jpg}
\caption{Heatmap of gene expression in cells grouped by annotations.}
\end{figure}

\hypertarget{vip-umaptsne}{%
\subsection{VIP -- UMAP/tSNE}\label{vip-umaptsne}}

To plot the embedding of cells in the selected group(s). One of pre-computed and loaded embeddings can be selected.

User can color cells in the embedding plots by multiple annotations (e.g., cell\_type, diagnosis).

Besides coloring cells by annotations, user can color cells based on gene expression level of selected genes in the embedding plots.

\begin{figure}
\centering
\includegraphics{figures/F10.jpg}
\caption{Embedding plotting of expression level of genes or gene set in the cells split by categories of an annotation}
\end{figure}

\hypertarget{vip-dot-plot}{%
\subsection{VIP -- Dot Plot}\label{vip-dot-plot}}

To show the fraction of cells (annotated by dot size) expressing a gene in each group and the averaged expression level of the gene (annotated by color intensity) in the group.

\begin{figure}
\centering
\includegraphics{figures/F11.jpg}
\caption{Dot plotting of the fraction of cells expressing genes above a cutoff in each categorie of the selected annotation}
\end{figure}

Note: The number of cells represented by side bar chart are always numbers of cells distributed in each category of certain annotation without filtering. It will give accurate estimate of number of cells in each bubble in the plot. The use of the plot is only meaningful when the counts matrix contains zeros representing no gene counts. Its visualization does not work for scaled or corrected matrices in which zero counts had been replaced by other values, see \url{https://scanpy-tutorials.readthedocs.io/en/multiomics/visualizing-marker-genes.html\#Dot-plots}.

\hypertarget{vip-track-plot}{%
\subsection{VIP -- Track Plot}\label{vip-track-plot}}

To show the expression of gene(s) of individual cells as vertical lines grouped by the selected annotation on x-axis. Instead of a color scale, the gene expression is represented by height.

\begin{figure}
\centering
\includegraphics{figures/F12.jpg}
\caption{Track plotting of expression of genes or gene set in each category of the selected annotation. Gene expression levels are represented by the heights of vertical lines}
\end{figure}

\hypertarget{vip-density-plot}{%
\subsection{VIP -- Density Plot}\label{vip-density-plot}}

To show the density of gene(s) expression in the cells annotated by category in the selected group(s) of cells. A density plot is a representation of the distribution of a numeric variable. It uses a kernel density estimate to show the probability density function of the variable (see more). It is a smoothed version of the histogram and is used in the same concept.

The bandwidth defines how close to a value point the distance between two points must be to influence the estimation of the density at the point. A small bandwidth only considers the closest values, so the estimation is close to the data. A large bandwidth considers more points and gives a smoother estimation.

\begin{figure}
\centering
\includegraphics{figures/F13.jpg}
\caption{Density plotting of the expression of genes in each group split by one annotation while colored by another.}
\end{figure}

\hypertarget{vip-2d-density-plot}{%
\subsection{VIP -- 2D Density Plot}\label{vip-2d-density-plot}}

Besides plotting of expression density of single gene, 2D density plot allows to explore the joint expression density of two genes in the cells expressing both genes above a cutoff.

\begin{figure}
\centering
\includegraphics{figures/F14.jpg}
\caption{2D Density plotting of expression of two genes in the selected cells}
\end{figure}

\hypertarget{vip-dual-genes}{%
\subsection{VIP -- Dual Genes}\label{vip-dual-genes}}

To view the relationship of expression levels of two genes in selected cells. It is based on the embedding plot of cells while coloring cells according to the expression levels of gene(s) in each cell.

\begin{figure}
\centering
\includegraphics{figures/F15.jpg}
\caption{Embedding plotting of the expression of two genes in the selected cell group(s)}
\end{figure}

\hypertarget{vip-sankey-diagram}{%
\subsection{VIP -- Sankey Diagram}\label{vip-sankey-diagram}}

Sankey diagram shows the flow of gene expression and annotations linked by cells. Gene expression is divided equally into bins so user can view distribution relationship between gene expression and annotations.

The diagram is also shown in an interactive way that user can change the layout by selecting several items (e.g., thin or thick on color bar, small or large space) from the panel. Also, user can drag these small boxes on the plot to get preferred layout and save it as high resolution SVG figure.

In addition, when you hover over mouse on a box, you can get detailed information about the source and target of flow

\begin{figure}
\centering
\includegraphics{figures/F16.jpg}
\caption{Sankey diagram provides quick and easy way to explore the inter-dependent relationship of variables.}
\end{figure}

\hypertarget{vip-stacked-barplot}{%
\subsection{VIP -- Stacked Barplot}\label{vip-stacked-barplot}}

To show the distribution of cells among categories of an annotation and/or ranges of expression of agene. Only two factors from annotations or genes can be chosen. The plot allows user to explore the distribution of cells in different views interactively.

\begin{figure}
\centering
\includegraphics{figures/F17A.jpg}
\caption{The distribution of cells in the selected group(s) regarding categories of an annotation and expression ranges of a gene}
\end{figure}

\begin{figure}
\centering
\includegraphics{figures/F17B.jpg}
\caption{Expand view to show percentage instead of numbers of cells}
\end{figure}

\begin{figure}
\centering
\includegraphics{figures/F17C.jpg}
\caption{Mouse over a bar to show details. Y-axis is re-located for better view and the number of cells represented by the bar is also shown in legend.}
\end{figure}

\hypertarget{vip-gene-detected}{%
\subsection{VIP -- Gene Detected}\label{vip-gene-detected}}

To show the number of genes expressed above the specified expression cut-off in the selected group(s) of cells.

\begin{figure}
\centering
\includegraphics{figures/F18.jpg}
\caption{The number of genes with expression over the cut-off in the cells from the selected group(s) of cells}
\end{figure}

\hypertarget{vip-deg-differential-expressed-genes}{%
\subsection{VIP -- DEG (Differential Expressed Genes)}\label{vip-deg-differential-expressed-genes}}

Besides plotting functions, cellxgene VIP also provides differential analysis between two selected groups of cells to identify differential expressed genes.

Three differential analysis statistical test methods are provided including Welch's t-test, Wilcoxon rank test and Wald's test. The statistical test results are presented in a table format including log2 Fold change, p-value and padj value (i.e, FDR value). Please note, we provide users with simple test methods for quick exploration within the interactive framework. However, there would be covariates need to be considered in a proper statistical test. Please consult your stats experts for appropriate test by using the right test method and right model.

Volcano plotting is also provided to show the log2FC vs.~-log10(FDR) relationship for all genes. User can select the gene(s) from the pre-selected gene list to be highlighted with text in the volcano plot.

\begin{figure}
\centering
\includegraphics{figures/F19A.jpg}
\caption{DEG analysis between the selected group(s)}
\end{figure}

\begin{figure}
\centering
\includegraphics{figures/F19B.jpg}
\caption{Volcano plotting of identified DEGs.}
\end{figure}

Note: The data used by DEG is unscaled (please refer to description of the dataset to find out what preprocessing was done on the data). Scaling control in the Figure Option does not apply to DEG. The three methods: `Welch's t-test' uses t-test (assuming underlining data with normal distributions) this uses cellxgene t-test implementation, `Wilcoxon rank test' uses Wilcoxon rank-sum test (does not assume known distributions, non-parametric test) and `Wald's test' uses Wald Chi-Squared test which is based on maximum likelihood. `Wilcoxon rank test' and `Wald's test' use diffxpy's implementation.

\hypertarget{vip-marker-genes}{%
\subsection{VIP -- Marker Genes}\label{vip-marker-genes}}

This functional module allows user to identify marker genes in the selected group(s) (more than 2, if 2 groups, please use DEG) of cells by annotation categories.

Four methods are provided for detecting marker genes including logreg, t-test, Wilcoxon, and t-testoverest\_var. For each identified marker gene, the gene name, scores (the z-score underlying the computation of a p-value for each gene for each group) and assigned group are listed in the output table.

In each annotation category, top ranked marker genes (this example shows top 2) will be plotted by score in comparison to the rest of the categories.

\begin{figure}
\centering
\includegraphics{figures/F20A.jpg}
\caption{Marker genes detection in the selected group(s) of cells regarding to the selected annotation categories}
\end{figure}

\begin{figure}
\centering
\includegraphics{figures/F20B.jpg}
\caption{Plotting of identified marker genes by each category}
\end{figure}

Note: The four methods implementations by calling scanpy.tl.rank\_genes\_groups function: `logreg' uses logistic regression, `t-test' uses t-test, `wilcoxon' uses Wilcoxon rank-sum, and `t-test\_overestim\_var' overestimates variance of each group.

\hypertarget{vip-command-line-interface}{%
\subsection{VIP -- Command Line Interface}\label{vip-command-line-interface}}

Although cellxgene VIP provides a rich set of visualization modules as shown above, command line interface is also built to allow unlimited visualization and analytical capabilities by power user who know how to program in Python / R languages.

\begin{figure}
\centering
\includegraphics{figures/F21.jpg}
\caption{Command line interface for user to program for advanced plotting and statistical analysis.}
\end{figure}

Note: In CLI the AnnData (adata) object is available by default, and it is processed as `Description' of the dataset states (i.e.: normalized and log transformed, but no scaled etc.). Settings in `Figure Option' tab won't apply to CLI.

\hypertarget{vip-comb.-abbr.}{%
\subsection{VIP -- Comb. \& Abbr.}\label{vip-comb.-abbr.}}

User can combine multiple annotations to create combinatorial annotation to group cells in various of plotting, e.g., stacked violin and dot plot. Firstly, user `uncheck all annotations' and, secondly go to the annotation panel in main window to select annotation categories to be combined, e.g., diagnosis (Control and MS) combined with sex (female and male). After clicking on `Create' button, all possible combinatorial names will be listed and `Custom\_combine' will be automatically available as an option in `Group by' drop down menu of many plotting functions.

User can also rename each annotation by creating abbreviations to shorten axis labels in figures.

\begin{figure}
\centering
\includegraphics{figures/F22.jpg}
\caption{Comb. \& Abbr. function allows user to create new annotation by combining multiple annotations and abbreviations to shorten axis labels in figures especially when custom combinatorial names are used}
\end{figure}

\hypertarget{vip-other-functions}{%
\subsection{VIP -- Other Functions}\label{vip-other-functions}}

There are other convenient functions available to user, such as `Save' or `Load' session, `Check All Annotations' and `Brush'.

`Save' or `Load' session are used to save the current cell selections and parameter settings to text file or load previously saved session file in the tool for visualization.

`Check All Annotations' is used to check all categorical selection boxes of annotations on the left panel.

`Brush' is to display exactly these selected ranges from histograms of variables on the right panel in a nice table that is not available in original cellxgene.

\begin{figure}
\centering
\includegraphics{figures/F23.jpg}
\caption{Other functions allow user to `Save' or `Load' session information, `Check All Annotations' and show values of brushed ranges on histograms}
\end{figure}

\hypertarget{methods}{%
\section{Methods}\label{methods}}

\hypertarget{client-side-integration-by-a-jspanel-window-vip}{%
\subsection*{Client-side Integration by a jsPanel Window (VIP)}\label{client-side-integration-by-a-jspanel-window-vip}}
\addcontentsline{toc}{subsection}{Client-side Integration by a jsPanel Window (VIP)}

Following section in config.sh file.

\begin{Shaded}
\begin{Highlighting}[]
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{"static/jquery.min.js"}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{link }\VariableTok{href=}\StringTok{\textquotesingle{}static/jspanel/dist/jspanel.css\textquotesingle{}} \VariableTok{rel=}\StringTok{\textquotesingle{}stylesheet\textquotesingle{}}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/jspanel.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/extensions/modal/jspanel.modal.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/extensions/tooltip/jspanel.tooltip.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/extensions/hint/jspanel.hint.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/extensions/layout/jspanel.layout.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/extensions/contextmenu/jspanel.contextmenu.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\OperatorTok{\textless{}}\NormalTok{script }\VariableTok{src=}\StringTok{\textquotesingle{}static/jspanel/dist/extensions/dock/jspanel.dock.js\textquotesingle{}}\OperatorTok{\textgreater{}\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\OperatorTok{\textless{}}\NormalTok{script}\OperatorTok{\textgreater{}}
  \CommentTok{// execute JavaScript code in panel content}
  \KeywordTok{var}\NormalTok{ setInnerHTML }\OperatorTok{=} \KeywordTok{function}\NormalTok{(elm}\OperatorTok{,}\NormalTok{ html) \{}
\NormalTok{    elm}\OperatorTok{.}\AttributeTok{innerHTML} \OperatorTok{=}\NormalTok{ html}\OperatorTok{;}
    \BuiltInTok{Array}\OperatorTok{.}\FunctionTok{from}\NormalTok{(elm}\OperatorTok{.}\FunctionTok{querySelectorAll}\NormalTok{(}\StringTok{\textquotesingle{}script\textquotesingle{}}\NormalTok{))}\OperatorTok{.}\FunctionTok{forEach}\NormalTok{( oldScript }\KeywordTok{=\textgreater{}}\NormalTok{ \{}
      \KeywordTok{const}\NormalTok{ newScript }\OperatorTok{=} \BuiltInTok{document}\OperatorTok{.}\FunctionTok{createElement}\NormalTok{(}\StringTok{\textquotesingle{}script\textquotesingle{}}\NormalTok{)}\OperatorTok{;}
      \BuiltInTok{Array}\OperatorTok{.}\FunctionTok{from}\NormalTok{(oldScript}\OperatorTok{.}\AttributeTok{attributes}\NormalTok{)}
      \OperatorTok{.}\FunctionTok{forEach}\NormalTok{( attr }\KeywordTok{=\textgreater{}}\NormalTok{ newScript}\OperatorTok{.}\FunctionTok{setAttribute}\NormalTok{(attr}\OperatorTok{.}\AttributeTok{name}\OperatorTok{,}\NormalTok{ attr}\OperatorTok{.}\AttributeTok{value}\NormalTok{) )}\OperatorTok{;}
\NormalTok{      newScript}\OperatorTok{.}\FunctionTok{appendChild}\NormalTok{(}\BuiltInTok{document}\OperatorTok{.}\FunctionTok{createTextNode}\NormalTok{(oldScript}\OperatorTok{.}\AttributeTok{innerHTML}\NormalTok{))}\OperatorTok{;}
\NormalTok{      oldScript}\OperatorTok{.}\AttributeTok{parentNode}\OperatorTok{.}\FunctionTok{replaceChild}\NormalTok{(newScript}\OperatorTok{,}\NormalTok{ oldScript)}\OperatorTok{;}
\NormalTok{    \})}\OperatorTok{;}
\NormalTok{  \}}
  \KeywordTok{var}\NormalTok{ plotPanel }\OperatorTok{=}\NormalTok{ jsPanel}\OperatorTok{.}\FunctionTok{create}\NormalTok{(\{}
    \DataTypeTok{panelSize}\OperatorTok{:} \StringTok{\textquotesingle{}190 0\textquotesingle{}}\OperatorTok{,}
    \DataTypeTok{position}\OperatorTok{:} \StringTok{\textquotesingle{}left{-}top 160 6\textquotesingle{}}\OperatorTok{,}
    \DataTypeTok{dragit}\OperatorTok{:}\NormalTok{ \{ }\DataTypeTok{containment}\OperatorTok{:}\NormalTok{ [}\OperatorTok{{-}}\DecValTok{10}\OperatorTok{,} \OperatorTok{{-}}\DecValTok{2000}\OperatorTok{,} \OperatorTok{{-}}\DecValTok{4000}\OperatorTok{,} \OperatorTok{{-}}\DecValTok{2000}\NormalTok{] \}}\OperatorTok{,} \CommentTok{// set dragging range of VIP window}
    \DataTypeTok{boxShadow}\OperatorTok{:} \DecValTok{1}\OperatorTok{,}
    \DataTypeTok{border}\OperatorTok{:} \StringTok{"solid \#D4DBDE thin"}\OperatorTok{,}
    \DataTypeTok{contentOverflow}\OperatorTok{:} \StringTok{\textquotesingle{}scroll scroll\textquotesingle{}}\OperatorTok{,} \CommentTok{// adding scrolling bars}
    \DataTypeTok{headerControls}\OperatorTok{:}\NormalTok{\{}
      \DataTypeTok{close}\OperatorTok{:} \StringTok{\textquotesingle{}remove\textquotesingle{}}\OperatorTok{,}
      \DataTypeTok{minimize}\OperatorTok{:} \StringTok{\textquotesingle{}remove\textquotesingle{}}\OperatorTok{,}
      \DataTypeTok{maximize}\OperatorTok{:} \StringTok{\textquotesingle{}remove\textquotesingle{}}
\NormalTok{    \}}\OperatorTok{,}
    \DataTypeTok{headerTitle}\OperatorTok{:} \KeywordTok{function}\NormalTok{ () \{}\ControlFlowTok{return} \StringTok{\textquotesingle{}\textless{}strong\textgreater{}Visualization in Plugin\textless{}/strong\textgreater{}\textquotesingle{}}\NormalTok{\}}\OperatorTok{,}
    \DataTypeTok{contentAjax}\OperatorTok{:}\NormalTok{ \{}
      \DataTypeTok{url}\OperatorTok{:} \StringTok{\textquotesingle{}static/interface.html\textquotesingle{}}\OperatorTok{,}
      \DataTypeTok{done}\OperatorTok{:} \KeywordTok{function}\NormalTok{ (panel) \{}
            \FunctionTok{setInnerHTML}\NormalTok{(panel}\OperatorTok{.}\AttributeTok{content}\OperatorTok{,} \KeywordTok{this}\OperatorTok{.}\AttributeTok{responseText}\NormalTok{)}\OperatorTok{;}
\NormalTok{      \}}
\NormalTok{    \}}\OperatorTok{,}
    \DataTypeTok{onwindowresize}\OperatorTok{:} \KeywordTok{function}\NormalTok{(}\BuiltInTok{event}\OperatorTok{,}\NormalTok{ panel) \{}
      \KeywordTok{var}\NormalTok{ jptop }\OperatorTok{=} \PreprocessorTok{parseInt}\NormalTok{(}\KeywordTok{this}\OperatorTok{.}\AttributeTok{currentData}\OperatorTok{.}\AttributeTok{top}\NormalTok{)}\OperatorTok{;}
      \KeywordTok{var}\NormalTok{ jpleft }\OperatorTok{=} \PreprocessorTok{parseInt}\NormalTok{(}\KeywordTok{this}\OperatorTok{.}\AttributeTok{currentData}\OperatorTok{.}\AttributeTok{left}\NormalTok{)}\OperatorTok{;}
      
      \ControlFlowTok{if}\NormalTok{ (jptop}\OperatorTok{\textless{}{-}}\DecValTok{10} \OperatorTok{||} \BuiltInTok{window}\OperatorTok{.}\AttributeTok{innerHeight}\OperatorTok{{-}}\NormalTok{jptop}\OperatorTok{\textless{}}\DecValTok{10} \OperatorTok{||} \BuiltInTok{window}\OperatorTok{.}\AttributeTok{innerWidth}\OperatorTok{{-}}\NormalTok{jpleft}\OperatorTok{\textless{}}\DecValTok{10} \OperatorTok{||}
\NormalTok{      jpleft}\OperatorTok{+}\PreprocessorTok{parseInt}\NormalTok{(}\KeywordTok{this}\OperatorTok{.}\AttributeTok{currentData}\OperatorTok{.}\AttributeTok{width}\NormalTok{)}\OperatorTok{\textless{}}\DecValTok{10}\NormalTok{) \{}
        \KeywordTok{this}\OperatorTok{.}\FunctionTok{reposition}\NormalTok{(}\StringTok{"left{-}top 160 6"}\NormalTok{)}\OperatorTok{;}
\NormalTok{      \}}
\NormalTok{    \}}\OperatorTok{,}
    \DataTypeTok{onunsmallified}\OperatorTok{:} \KeywordTok{function}\NormalTok{ (panel}\OperatorTok{,}\NormalTok{ status) \{}
      \KeywordTok{this}\OperatorTok{.}\FunctionTok{reposition}\NormalTok{(}\StringTok{\textquotesingle{}center{-}top {-}370 180\textquotesingle{}}\NormalTok{)}\OperatorTok{;}
      \KeywordTok{this}\OperatorTok{.}\FunctionTok{resize}\NormalTok{(\{ }\DataTypeTok{width}\OperatorTok{:} \DecValTok{740}\OperatorTok{,} \DataTypeTok{height}\OperatorTok{:} \KeywordTok{function}\NormalTok{() \{ }\ControlFlowTok{return} \BuiltInTok{Math}\OperatorTok{.}\FunctionTok{min}\NormalTok{(}\DecValTok{480}\OperatorTok{,} \BuiltInTok{window}\OperatorTok{.}\AttributeTok{innerHeight}\OperatorTok{*}\FloatTok{0.6}\NormalTok{)}\OperatorTok{;}\NormalTok{\} \})}\OperatorTok{;}
\NormalTok{    \}}\OperatorTok{,}
    \DataTypeTok{onsmallified}\OperatorTok{:} \KeywordTok{function}\NormalTok{ (panel}\OperatorTok{,}\NormalTok{ status) \{}
      \KeywordTok{this}\OperatorTok{.}\FunctionTok{reposition}\NormalTok{(}\StringTok{\textquotesingle{}left{-}top 160 6\textquotesingle{}}\NormalTok{)}\OperatorTok{;}
      \KeywordTok{this}\OperatorTok{.}\AttributeTok{style}\OperatorTok{.}\AttributeTok{width} \OperatorTok{=} \StringTok{\textquotesingle{}190px\textquotesingle{}}\OperatorTok{;}
\NormalTok{    \}}
\NormalTok{  \})}\OperatorTok{.}\FunctionTok{smallify}\NormalTok{()}\OperatorTok{;}
\NormalTok{  plotPanel}\OperatorTok{.}\AttributeTok{headerbar}\OperatorTok{.}\AttributeTok{style}\OperatorTok{.}\AttributeTok{background} \OperatorTok{=} \StringTok{"\#D4DBDE"}\OperatorTok{;}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\OperatorTok{\textless{}}\NormalTok{/script}\OperatorTok{\textgreater{}}
\ExtensionTok{EOF}
\VariableTok{insertL=$(}\FunctionTok{sed} \AttributeTok{{-}e} \StringTok{\textquotesingle{}s/[\&\textbackslash{}\textbackslash{}/]/\textbackslash{}\textbackslash{}\&/g; s/|/\textbackslash{}\textbackslash{}|/g; s/$/\textbackslash{}\textbackslash{}/;\textquotesingle{}} \AttributeTok{{-}e} \StringTok{\textquotesingle{}$s/\textbackslash{}\textbackslash{}$//\textquotesingle{}} \OperatorTok{\textless{}\textless{}\textless{}}\StringTok{"}\VariableTok{$insertL}\StringTok{"}\VariableTok{)}
\FunctionTok{sed} \AttributeTok{{-}i} \StringTok{"s|\textless{}div id=}\DataTypeTok{\textbackslash{}"}\StringTok{root}\DataTypeTok{\textbackslash{}"}\StringTok{\textgreater{}\textless{}/div\textgreater{}|}\VariableTok{$insertL}\StringTok{\textbackslash{}n\&|"} \StringTok{"cellxgene/client/index\_template.html"}
\end{Highlighting}
\end{Shaded}

All functional VIP HTML and JavaScript code will be in ``interface.html'' that is independent of cellxgene code bases.

\hypertarget{server-side-integration}{%
\subsection*{Server-side Integration}\label{server-side-integration}}
\addcontentsline{toc}{subsection}{Server-side Integration}

Following section in config.sh file.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{echo} \StringTok{\textquotesingle{}}
\StringTok{from server.app.biogenInterface import route  }
\StringTok{@webbp.route("/biogen", methods=["POST"]) }
\StringTok{def biogen():}
\StringTok{  return route(request.data,current\_app.app\_config)\textquotesingle{}} \OperatorTok{\textgreater{}\textgreater{}}\NormalTok{ cellxgene/server/app/app.py}
\BuiltInTok{.}
\BuiltInTok{.}
\BuiltInTok{.}

\VariableTok{strPath=}\StringTok{"}\VariableTok{$(}\ExtensionTok{python} \AttributeTok{{-}c} \StringTok{\textquotesingle{}import site; print(site.getsitepackages())\textquotesingle{}}\VariableTok{)}\StringTok{"}
\VariableTok{strPath=$\{strPath}\OperatorTok{//}\StringTok{"[\textquotesingle{}"}\OperatorTok{/}\VariableTok{\}}
\VariableTok{strPath=$\{strPath}\OperatorTok{//}\StringTok{"\textquotesingle{}]"}\OperatorTok{/}\VariableTok{\}}
\VariableTok{strweb=}\StringTok{"}\VariableTok{$\{strPath\}}\StringTok{/server/common/web/static/."}
\BuiltInTok{echo} \VariableTok{$strweb}
\FunctionTok{cp}\NormalTok{ interface.html }\VariableTok{$strweb}
\FunctionTok{cp}\NormalTok{ jquery.min.js }\VariableTok{$strweb}
\FunctionTok{cp}\NormalTok{ color\_map.png }\VariableTok{$strweb}

\FunctionTok{cp} \AttributeTok{{-}R}\NormalTok{ DataTables }\VariableTok{$strweb}
\FunctionTok{cp} \AttributeTok{{-}R}\NormalTok{ jspanel }\VariableTok{$strweb}

\FunctionTok{cp}\NormalTok{ cellxgene/server/test/decode\_fbs.py }\VariableTok{$strPath}\NormalTok{/server/app/.}
\FunctionTok{cp}\NormalTok{ VIPInterface.py }\VariableTok{$strPath}\NormalTok{/server/app/.}
\end{Highlighting}
\end{Shaded}

\hypertarget{communication-between-vip-and-cellxgene-web-gui}{%
\subsection*{Communication between VIP and cellxgene web GUI}\label{communication-between-vip-and-cellxgene-web-gui}}
\addcontentsline{toc}{subsection}{Communication between VIP and cellxgene web GUI}

Cellxgene client utilizes React Redux that is the official React binding for Redux. It lets your React components read data from a Redux store, and dispatch actions to the store to update data.

So, this line of code is appended to the end of client/src/reducers/index.js of cellxgene source code to expose the store to the browser.

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{window}\OperatorTok{.}\AttributeTok{store} \OperatorTok{=}\NormalTok{ store}\OperatorTok{;}
\end{Highlighting}
\end{Shaded}

By doing this, Redux store holding client data and user selections are visible to VIP to access variables and dispatch actions to control cellxgene user interface. For example,

\begin{itemize}
\tightlist
\item
  Unselect / select a feature. GUI is refreshed automatically after dispatching.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\BuiltInTok{window}\OperatorTok{.}\AttributeTok{store}\OperatorTok{.}\FunctionTok{dispatch}\NormalTok{(\{}\DataTypeTok{type}\OperatorTok{:} \StringTok{"categorical metadata filter deselect"}\OperatorTok{,} \DataTypeTok{metadataField}\OperatorTok{:} \StringTok{"louvain"}\OperatorTok{,} \DataTypeTok{categoryIndex}\OperatorTok{:} \DecValTok{5}\NormalTok{\})}
\BuiltInTok{window}\OperatorTok{.}\AttributeTok{store}\OperatorTok{.}\FunctionTok{dispatch}\NormalTok{(\{}\DataTypeTok{type}\OperatorTok{:} \StringTok{"categorical metadata filter select"}\OperatorTok{,} \DataTypeTok{metadataField}\OperatorTok{:} \StringTok{"louvain"}\OperatorTok{,} \DataTypeTok{categoryIndex}\OperatorTok{:} \DecValTok{5}\NormalTok{\})}
\end{Highlighting}
\end{Shaded}

\begin{itemize}
\tightlist
\item
  Get state of just finished action and synchronize gene input and cell selections from main window to VIP if corresponding action was performed.
\end{itemize}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{const}\NormalTok{ unsubscribe }\OperatorTok{=} \BuiltInTok{window}\OperatorTok{.}\AttributeTok{store}\OperatorTok{.}\FunctionTok{subscribe}\NormalTok{(() }\KeywordTok{=\textgreater{}}\NormalTok{ \{}
  \ControlFlowTok{if}\NormalTok{ (}\BuiltInTok{window}\OperatorTok{.}\AttributeTok{store}\OperatorTok{.}\FunctionTok{getState}\NormalTok{()[}\StringTok{"@@undoable/filterState"}\NormalTok{]}\OperatorTok{.}\AttributeTok{prevAction}\NormalTok{) \{}
\NormalTok{    actionType }\OperatorTok{=} \BuiltInTok{window}\OperatorTok{.}\AttributeTok{store}\OperatorTok{.}\FunctionTok{getState}\NormalTok{()[}\StringTok{"@@undoable/filterState"}\NormalTok{]}\OperatorTok{.}\AttributeTok{prevAction}\OperatorTok{.}\AttributeTok{type}\OperatorTok{;}
    \ControlFlowTok{if}\NormalTok{ (actionType}\OperatorTok{.}\FunctionTok{includes}\NormalTok{(}\StringTok{"user defined gene success"}\NormalTok{) }\OperatorTok{||}
\NormalTok{    actionType}\OperatorTok{.}\FunctionTok{includes}\NormalTok{(}\StringTok{"store current cell selection as differential set"}\NormalTok{)) \{}
      \FunctionTok{sync}\NormalTok{()}\OperatorTok{;}
\NormalTok{      \}}
\NormalTok{  \}}
\NormalTok{\})}\OperatorTok{;}
\end{Highlighting}
\end{Shaded}

\hypertarget{diffxpy-integration}{%
\subsection*{Diffxpy Integration}\label{diffxpy-integration}}
\addcontentsline{toc}{subsection}{Diffxpy Integration}

This is the sample pseudocode, please see VIPInterface.py for actual implementation.

\begin{Shaded}
\begin{Highlighting}[]
\ImportTok{import}\NormalTok{ scanpy }\ImportTok{as}\NormalTok{ sc}
\ImportTok{import}\NormalTok{ pandas }\ImportTok{as}\NormalTok{ pd}
\ImportTok{import}\NormalTok{ diffxpy.api }\ImportTok{as}\NormalTok{ app}
\CommentTok{\# set 1 of cells as cell1; set 2 of cells as cell2}


\ControlFlowTok{with}\NormalTok{ app.get\_data\_adaptor() }\ImportTok{as}\NormalTok{ data\_adaptor:}
\NormalTok{  X1 }\OperatorTok{=}\NormalTok{ data\_adaptor.data.X[cell1]}
\NormalTok{  X2 }\OperatorTok{=}\NormalTok{ data\_adaptor.data.X[cell2]}


\NormalTok{adata }\OperatorTok{=}\NormalTok{ sc.AnnData(pd.concat([X1,X2]),pd.DataFrame([}\StringTok{\textquotesingle{}grp1\textquotesingle{}} \ControlFlowTok{for}\NormalTok{ i }\KeywordTok{in} \BuiltInTok{range}\NormalTok{(X1.shape[}\DecValTok{0}\NormalTok{])]}\OperatorTok{+}\NormalTok{[}\StringTok{\textquotesingle{}grp2\textquotesingle{}} \ControlFlowTok{for}\NormalTok{ i }\KeywordTok{in} \BuiltInTok{range}\NormalTok{(X2.shape[}\DecValTok{0}\NormalTok{])],columns}\OperatorTok{=}\NormalTok{[}\StringTok{\textquotesingle{}comGrp\textquotesingle{}}\NormalTok{]))}
\NormalTok{deg }\OperatorTok{=}\NormalTok{ de.test.two\_sample(adata,}\StringTok{\textquotesingle{}comGrp\textquotesingle{}}\NormalTok{).summary()}
\CommentTok{\#deg is a dataframe contains the folloing columns [\textquotesingle{}gene\textquotesingle{},\textquotesingle{}log2fc\textquotesingle{},\textquotesingle{}pval\textquotesingle{},\textquotesingle{}qval\textquotesingle{}]}
\end{Highlighting}
\end{Shaded}

\hypertarget{create-h5ad-file-from-seurat-object}{%
\subsection{Create h5ad file from Seurat object}\label{create-h5ad-file-from-seurat-object}}

First, export the following from Seurat object in R: \textbf{expression matrix (assume normalized), metadata and coordinates (pca, tsne, umap) as separate txt files.}

Next in Python, create an AnnData object from 10x (scanpy.read\_h5ad function) as a starting point. Then replace the expression matrix, meta data and coordinates as following, a h5ad file would be generated.

\begin{Shaded}
\begin{Highlighting}[]
\ImportTok{import}\NormalTok{ sys}
\ImportTok{import}\NormalTok{ scanpy }\ImportTok{as}\NormalTok{ sc}
\ImportTok{import}\NormalTok{ pandas }\ImportTok{as}\NormalTok{ pd}
\ImportTok{import}\NormalTok{ numpy }\ImportTok{as}\NormalTok{ np}
\ImportTok{import}\NormalTok{ seaborn }\ImportTok{as}\NormalTok{ sns}
\ImportTok{from}\NormalTok{ numpy }\ImportTok{import}\NormalTok{ ndarray, unique}
\ImportTok{from}\NormalTok{ scipy.sparse.csc }\ImportTok{import}\NormalTok{ csc\_matrix}

\NormalTok{adata}\OperatorTok{=}\NormalTok{ sc.read\_h5ad(}\StringTok{"previous generated .h5ad"}\NormalTok{)}

\CommentTok{\# read clustering res}
\NormalTok{xpca }\OperatorTok{=}\NormalTok{ pd.read\_csv(“.}\OperatorTok{/}\NormalTok{data}\OperatorTok{/}\NormalTok{harmony\_clustered.h5ad.pca\_coordinates.txt}\StringTok{", sep=\textquotesingle{}}\CharTok{\textbackslash{}t}\StringTok{\textquotesingle{}, encoding=\textquotesingle{}utf{-}8\textquotesingle{})}
\StringTok{xtsne = pd.read\_csv(“./data/harmony\_clustered.h5ad.tsne\_coordinates.txt"}\NormalTok{, sep}\OperatorTok{=}\StringTok{\textquotesingle{}}\CharTok{\textbackslash{}t}\StringTok{\textquotesingle{}}\NormalTok{, encoding}\OperatorTok{=}\StringTok{\textquotesingle{}utf{-}8\textquotesingle{}}\NormalTok{)}
\NormalTok{xumap }\OperatorTok{=}\NormalTok{ pd.read\_csv(“.}\OperatorTok{/}\NormalTok{data}\OperatorTok{/}\NormalTok{harmony\_clustered.h5ad.umap\_coordinates.txt}\StringTok{", sep=\textquotesingle{}}\CharTok{\textbackslash{}t}\StringTok{\textquotesingle{}, encoding=\textquotesingle{}utf{-}8\textquotesingle{})}
\StringTok{xobs = pd.read\_csv(“./data/harmony\_clustered.h5ad.meta\_data.txt"}\NormalTok{, sep}\OperatorTok{=}\StringTok{\textquotesingle{}}\CharTok{\textbackslash{}t}\StringTok{\textquotesingle{}}\NormalTok{, encoding}\OperatorTok{=}\StringTok{\textquotesingle{}utf{-}8\textquotesingle{}}\NormalTok{)}

\NormalTok{xpca.set\_index(}\StringTok{\textquotesingle{}index\textquotesingle{}}\NormalTok{, inplace}\OperatorTok{=}\VariableTok{True}\NormalTok{)}
\NormalTok{xtsne.set\_index(}\StringTok{\textquotesingle{}index\textquotesingle{}}\NormalTok{, inplace}\OperatorTok{=}\VariableTok{True}\NormalTok{)}
\NormalTok{xumap.set\_index(}\StringTok{\textquotesingle{}index\textquotesingle{}}\NormalTok{, inplace}\OperatorTok{=}\VariableTok{True}\NormalTok{)}
\NormalTok{xobs.set\_index(}\StringTok{\textquotesingle{}index\textquotesingle{}}\NormalTok{, inplace}\OperatorTok{=}\VariableTok{True}\NormalTok{)}

\NormalTok{adata.obsm[}\StringTok{\textquotesingle{}X\_pca\textquotesingle{}}\NormalTok{] }\OperatorTok{=}\NormalTok{ np.array(xpca.loc[adataRaw.obs.index])}

\NormalTok{adata.obsm[}\StringTok{\textquotesingle{}X\_tsne\textquotesingle{}}\NormalTok{] }\OperatorTok{=}\NormalTok{ np.array(xtsne.loc[adataRaw.obs.index])}
\NormalTok{adata.obsm[}\StringTok{\textquotesingle{}X\_umap\textquotesingle{}}\NormalTok{] }\OperatorTok{=}\NormalTok{ np.array(xumap.loc[adataRaw.obs.index])}
\NormalTok{adata.obs }\OperatorTok{=}\NormalTok{ xobs.loc[adataRaw.obs.index] }\CommentTok{\# this is a pandas dataframe}

\CommentTok{\# read in expression matrix as numpy.ndarray as following:}
\NormalTok{exp\_mat }\OperatorTok{=}\NormalTok{ np.loadtxt(fname }\OperatorTok{=}\NormalTok{”expression matrix .txt}\StringTok{")}
\StringTok{adata.X = exp\_mat}

\StringTok{\# convert dense matrix into sparse matrix to save storage space and memory usage}
\StringTok{adata.X = csc\_matrix(adata.X)\_matrix}

\StringTok{\# add short description and initial graph settings. “|” and “by” are delimiters for VIP to parse the initial settings. Please follow the same rule for your own h5ad files.}
\StringTok{adata.obs[\textquotesingle{}\textgreater{}Description\textquotesingle{}] = [\textquotesingle{}Human brain snRNAseq 46k cells (MS Nature 2019 Schirmer et al.); data normalized, log transformed and scaled UMI; platform {-} 10X v2 chemistry | embedding by umap; color by cell\_type\textquotesingle{}]*adata.n\_obs}

\StringTok{\# Then last step to save h5ad:}
\StringTok{adata.write\_h5ad("}\NormalTok{final output.h5ad}\StringTok{")}
\end{Highlighting}
\end{Shaded}

\hypertarget{helpful-tips}{%
\section{Helpful Tips}\label{helpful-tips}}

\hypertarget{handle-nulls-in-categorical-annotation}{%
\subsection{Handle nulls in categorical annotation}\label{handle-nulls-in-categorical-annotation}}

Such nan's in categorical annotation would cause trouble in VIP because it cannot be converted to string. Here is how to handle it, let's call the annotation X\_annotation:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{\# Cast to str from categorical}
\NormalTok{adata.obs }\OperatorTok{=}\NormalTok{ adata.obs.astype(\{}\StringTok{\textquotesingle{}X\_annotation\textquotesingle{}}\NormalTok{:}\StringTok{\textquotesingle{}str\textquotesingle{}}\NormalTok{\})}

\CommentTok{\# replace all of nan by ‘nan’}
\NormalTok{adata.obs[}\StringTok{"X\_annotation "}\NormalTok{][adata.obs[}\StringTok{"X\_annotation "}\NormalTok{].isnull()] }\OperatorTok{=} \StringTok{\textquotesingle{}nan\textquotesingle{}}
\end{Highlighting}
\end{Shaded}

\hypertarget{display-full-traceback-stack-for-debugging-in-vip}{%
\subsection{Display full traceback stack for debugging in VIP}\label{display-full-traceback-stack-for-debugging-in-vip}}

It follows the global setting. Please set ``---verbose'' to launch cellxgene server.

\hypertarget{pitfall-of-using-special-characters}{%
\subsection{Pitfall of using special characters}\label{pitfall-of-using-special-characters}}

In the mode which allows user to create manual annotation in cellxgene, user should try to avoid using hyphen (``-'') in name label.It would cause client-side issue. Please try to use underscores.

\hypertarget{potential-use-for-bulk-or-pseudo-bulk-sample-dataset}{%
\subsection{Potential use for bulk or pseudo bulk sample dataset}\label{potential-use-for-bulk-or-pseudo-bulk-sample-dataset}}

Once the data matrix is replaced by sample x gene matrix, cellxgene VIP framework can handle regular bulk / pseudobulk RNAseq datasets. Simply replace ``cells'' by ``samples''. All plotting functions would still work.

\end{document}
